File: Firebase/Database/Core/View/FCacheNode.m
/*
 * Copyright 2017 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FCacheNode.h"
#import "FNode.h"
#import "FPath.h"
#import "FEmptyNode.h"
#import "FIndexedNode.h"

@interface FCacheNode ()
@property (nonatomic, readwrite) BOOL isFullyInitialized;
@property (nonatomic, readwrite) BOOL isFiltered;
@property (nonatomic, strong, readwrite) FIndexedNode *indexedNode;
@end

@implementation FCacheNode
- (id) initWithIndexedNode:(FIndexedNode *)indexedNode
        isFullyInitialized:(BOOL)fullyInitialized
                isFiltered:(BOOL)filtered
{
    self = [super init];
    if (self) {
        self.indexedNode = indexedNode;
        self.isFullyInitialized = fullyInitialized;
        self.isFiltered = filtered;
    }
    return self;
}

- (BOOL)isCompleteForPath:(FPath *)path {
    if (path.isEmpty) {
        return self.isFullyInitialized && !self.isFiltered;
    } else {
        NSString *childKey = [path getFront];
        return [self isCompleteForChild:childKey];
    }
}

- (BOOL)isCompleteForChild:(NSString *)childKey {
    return (self.isFullyInitialized && !self.isFiltered) || [self.node hasChild:childKey];
}

- (id<FNode>)node {
    return self.indexedNode.node;
}

@end

File: FirebaseDatabaseSwiftCore/Sources/FCacheNode.swift
//
//  File.swift
//  
//
//  Created by Morten Bek Ditlevsen on 19/02/2022.
//

import Foundation


@objc public class FCacheNode: NSObject {
  @objc public var isFullyInitialized: Bool
  @objc public var isFiltered: Bool
  public var indexedNode: FIndexedNode
    @objc(indexedNode) public var indexedNodeObjC: FIndexedNodeObjC {
        .init(wrapped: indexedNode)
    }
  @objc public var node: FNode {
    indexedNode.node
  }
  public init(indexedNode: FIndexedNode, isFullyInitialized: Bool, isFiltered: Bool) {
    self.indexedNode = indexedNode
    self.isFiltered = isFiltered
    self.isFullyInitialized = isFullyInitialized
  }

  @objc public func isComplete(forPath path: FPath) -> Bool {
    if let childKey = path.getFront() {
      return isComplete(forChild: childKey)
    } else { // path is empty
      return isFullyInitialized && !isFiltered
    }
  }

  @objc public func isComplete(forChild childKey: String) -> Bool {
    (isFullyInitialized && !isFiltered) || node.hasChild(childKey)
  }
}
